plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -443,6 +473,58 @@ private static <K, V> Map<K, V> getAll(Cache<K, AtomicReference<V>> cache, Itera
         return result.buildOrThrow();
     }
 
+    private static <K, V> Map<K, V> getAll(
+            Cache<K, AtomicReference<V>> cache,
+            Iterable<K> keys,
+            Function<Collection<K>, Map<K, V>> bulkLoader,
+            Predicate<V> shouldReload,
+            BiFunction<V, V, V> merger)
+    {
+        ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+        Map<K, AtomicReference<V>> valueHolders = new HashMap<>();
+
+        ImmutableList.Builder<K> missingKeysBuilder = ImmutableList.builder();
+        keys.forEach(key -> {
+            // make sure the value holder is retrieved before the new values are loaded
+            // so that in case of invalidation we will not set the stale value
+            AtomicReference<V> currentValueHolder = uncheckedCacheGet(cache, key, AtomicReference::new);
+            valueHolders.put(key, currentValueHolder);
+            V currentValue = currentValueHolder.get();
+            if (currentValue != null && !shouldReload.test(currentValue)) {
+                result.put(key, currentValue);
+            }
+            else {
+                missingKeysBuilder.add(key);
+            }
+        });
+
+        List<K> missingKeys = missingKeysBuilder.build();
+
+        if (missingKeys.isEmpty()) {
+            return result.buildOrThrow();
+        }
+
+        Map<K, V> newEntries = bulkLoader.apply(missingKeys);
+        newEntries.forEach((key, newValue) -> {
+            AtomicReference<V> valueHolder = valueHolders.get(key);
+            V oldValue = valueHolder.get();
+            if (newValue == null) {
+                throw new InvalidCacheLoadException("Failed to return a value for " + key);
+            }
+            V newResult;
+            if (oldValue == null) {
+                newResult = newValue;
+            }
+            else {
+                newResult = merger.apply(oldValue, newValue);
+            }
+            result.put(key, newResult);
+            valueHolder.set(newResult);
```

Comment:
compare and set? accumulateAndGet? 

we don't want to clobber "better" values written by other threads